hudi-spark-datasource/hudi-spark-common/src/main/java/org/apache/hudi/internal/BulkInsertDataInternalWriterHelper.java

```diff
@@ -157,7 +152,29 @@ public List<HoodieInternalWriteStatus> getWriteStatuses() throws IOException {
     return writeStatusList;
   }
 
-  public void abort() {
+  public void abort() {}
+
+  public void close() throws IOException {
+    for (HoodieRowCreateHandle rowCreateHandle : handles.values()) {
+      writeStatusList.add(rowCreateHandle.close());
+    }
+    handles.clear();
+    handle = null;
+  }
+
+  private UTF8String extractPartitionPath(InternalRow row) {
+    if (populateMetaFields) {
+      // In case meta-fields are materialized w/in the table itself, we can just simply extract
+      // partition path from there
+      //
+      // NOTE: Helper keeps track of [[lastKnownPartitionPath]] as [[UTF8String]] to avoid
+      //       conversion from Catalyst internal representation into a [[String]]
+      return row.getUTF8String(HoodieRecord.PARTITION_PATH_META_FIELD_POS);
+    } else if (keyGeneratorOpt.isPresent()) {
+      return keyGeneratorOpt.get().getPartitionPath(row, structType);
```

Comment:
Didn't miss it, it's actually one of the primary premises of this PR to get red of such special handling: special handling duplicates the logic across many places (key-gen, bulk-insert handler, etc), which surely enough leads to bugs induced by code getting out of sync (for ex, broken hive-style partitioning/encoding in bulk-insert path, etc).

Keys/partition-path should be produced by a single SoT which in this case is the KG, and it should be done in a performant (ie engine-specific) way, and that's exactly what this PR tackles.

Reply:
+1 on the SoT. the if clauses across the code base are hard to maintain. IIRC that was done to avoid UDF overhead in the first place? which may not be needed anymore?

That said, can you both confirm there are no perf side-effects to doing it either way?